### ISSUE TITLE
Throwing an error during a transaction doesn't allow the lock to roll back

### ISSUE THREAD

**Issue by OultimoCoder** [depot/kysely-planetscale#20]
I am using the latest version.

Here is my code:

```
export const createOauthUser = async (
  providerUser: OauthUser,
  databaseConfig: Config['database']
) => {
  const db = getDBClient(databaseConfig)
  try {
    await db.transaction().execute(async (trx) => {
      const userId = await trx
        .insertInto('user')
        .values({
          name: providerUser.name,
          email: providerUser.email,
          is_email_verified: true,
          password: null,
          role: 'user'
        })
        .executeTakeFirstOrThrow()
      console.log(userId)
      await trx
        .insertInto('authorisations')
        .values({
          user_id: Number(userId.insertId),
          provider_type: providerUser.providerType,
          provider_user_id: providerUser.id.toString()
        })
        .executeTakeFirstOrThrow()
      console.log('ended')
      return userId
    })
  } catch (error) {
    try {
      await db.deleteFrom('user').where('user.email', '=', providerUser.email).execute()
    } catch (err) {
      console.log(err)
    }
    console.log('here')
    throw new ApiError(
      httpStatus.FORBIDDEN,
      `Cannot signup with ${providerUser.providerType}, user already exists with that email`
    )
  }
  const user = await getUserByProviderIdType(
    providerUser.id.toString(), providerUser.providerType, databaseConfig
  )
  return User.convert(user)
}
```

I added the delete statement in the try catch purely to handle an error I am having. If I throw an error during my transaction it should rollback the transaction and release the locks but I instead get this error intermittently (the other times it completes in under 20 seconds):

```
DatabaseError: target: dictionary-api.-.primary: vttablet: rpc error: code = DeadlineExceeded desc = Lock wait timeout exceeded; try restarting transaction (errno 1205) (sqlstate HY000) (CallerID: ihkeoohzfw2oodprmc73): Sql: "delete from `user` where `user`.email = :user_email", BindVars: {REDACTED}
        at Connection.execute (/backend/node_modules/@planetscale/database/dist/index.js:78:19)
        at processTicksAndRejections (node:internal/process/task_queues:95:5)
        at _PlanetScaleConnection.executeQuery (/backend/node_modules/kysely-planetscale/dist/index.js:92:21)
        at /backend/node_modules/kysely/dist/esm/query-executor/query-executor-base.js:35:28
        at DefaultConnectionProvider.provideConnection (/backend/node_modules/kysely/dist/esm/driver/default-connection-provider.js:10:20)
        at DefaultQueryExecutor.executeQuery (/backend/node_modules/kysely/dist/esm/query-executor/query-executor-base.js:34:16)
        at DeleteQueryBuilder.execute (/backend/node_modules/kysely/dist/esm/query-builder/delete-query-builder.js:392:24)
        at Module.createOauthUser (/backend/src/services/user.service.ts:65:7)
        at Module.loginOrCreateUserWithOauth (/backend/src/services/auth.service.ts:93:19)
        at oauthCallback (/backend/src/controllers/auth/oauth/oauth.controller.ts:27:16) {
      status: 400,
      body: {
        message: 'target: dictionary-api.-.primary: vttablet: rpc error: code = DeadlineExceeded desc = Lock wait timeout exceeded; try restarting transaction (errno 1205) (sqlstate HY000) (CallerID: ihkeoohzfw2oodprmc73): Sql: "delete from `user` where `user`.email = :user_email", BindVars: {REDACTED}',
        code: 'UNKNOWN'
      }
    }
```

This is the sole test I am running and the sole code executing.

**Comment by OultimoCoder** [comment]
I found this upgrading from 1.1 to 1.3

**Comment by jacobwgillespie** [comment]
@OultimoCoder this used to work for you with version 1.1? I'm not seeing anything in the diff that would directly change the behavior of transactions: https://github.com/depot/kysely-planetscale/compare/v1.1.0...v1.3.0.

Did you upgrade anything else at the same time (`kysely` or `@planetscale/database`)? It still could be something here, but if not, the other thing to check would be if another request was holding a lock that conflicted with the `DELETE`, I'm not super familiar with what Vitess locks on delete.

**Comment by jacobwgillespie** [comment]
Interesting - are you setting the `useSharedConnection` config option? That's the thing that 1.2 added, but it's supposed to be disabled by default.

**Comment by OultimoCoder** [comment]
I haven't made any changes from 1.1, just upgraded the package to 1.3.

**Comment by OultimoCoder** [comment]
Could it be related to the changes made that fixed this other issue I reported with transactions in 1.3 #17

**Comment by jacobwgillespie** [comment]
Well v1.2.0 should have that fix included, so if you're not able to reproduce on 1.2, then it seems unrelated to that.

**Comment by OultimoCoder** [comment]
```
import { Kysely } from 'kysely'
import { PlanetScaleDialect } from 'kysely-planetscale'
import { ApiKeyTable } from '../models/apiKey.model'
import { AuthProviderTable } from '../models/authProvider.model'
import { EntriesTable } from '../models/entry.model'
import { FieldSourcesTable } from '../models/fieldSource.model'
import { SentencesTable } from '../models/sentence.model'
import { SourcesTable } from '../models/source.model'
import { UserTable } from '../models/user.model'
import { Config } from './config'

let dbClient: Kysely<Database>

export interface Database {
  user: UserTable,
  authorisations: AuthProviderTable
  api_keys: ApiKeyTable,
  sentences: SentencesTable,
  entries: EntriesTable,
  sources: SourcesTable,
  field_sources: FieldSourcesTable
}

export const getDBClient = (databaseConfig: Config['database']) => {
  dbClient =
    dbClient ||
    new Kysely<Database>({
      dialect: new PlanetScaleDialect({
        username: databaseConfig.username,
        password: databaseConfig.password,
        host: databaseConfig.host
      })
    })
  return dbClient
}
```

This is how my client is instantiated, I have done 0 code changes when testing upgrades.

**Comment by OultimoCoder** [comment]
And I did a repeat test just to check, 100% something introduced between 1.2 and 1.3 which causes this error.

**Comment by jacobwgillespie** [comment]
Oh actually 1.2.1 included the transaction fixes from #17, 1.2.0 did not. Did you try it with 1.2.1, or just 1.2.0?

**Comment by OultimoCoder** [comment]
Fails with 1.2.1!

**Comment by jacobwgillespie** [comment]
Yeah this is interesting, it's almost as if the `rollback` is not called? 1.2.1 updated the logic to `await` the rollback (and everything else), but I'm not sure how that would cause the issue you're seeing.

@koskimas do you see anything I'm doing incorrectly here? [`v1.2.0...v1.2.1`#diff-a2a171449d](https://github.com/depot/kysely-planetscale/compare/v1.2.0...v1.2.1#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80). I'm not super familiar with what Kysely internals do when an error is thrown inside a transaction.

**Comment by sromic** [comment]
@OultimoCoder @jacobwgillespie I was able to replicate this issue but only when I put delete statement into catch block within the transaction scope, but not like it's in the original code.

Now haven't find the root cause for this 20 sec query execution time so far, but there is something what could help @OultimoCoder to mitigate this problem.

Basically, I found that `v1.2.0` doesn't perform a rollbacks in case of an error within the transaction - I guess that's why in the given code there is a delete statement in case of the error.
Starting with `v1.2.1` this seems to be fixed, and rollbacks are performed - inserts aren't happening.

So with rollbacks truly supported now, @OultimoCoder you should be able to get rid of delete statement in catch block and have rollback do it for you.

**Comment by OultimoCoder** [comment]
I originally found this when running tests using jest and a delete test would run after a test that intentionally errored in the catch block.

The delete code in the catch block is just an easy way of replicating it.

My issue is the 20 second timeout with the locks not being released after an error in the catch block.

**Comment by jacobwgillespie** [comment]
I haven't had a chance to investigate further yet - the thing that was fixed in `v1.2.1` (#10) was that transaction commands (`BEGIN`, `COMMIT`, `ROLLBACK`) weren't actually being awaited, so they could fail silently.

What I don't know here is if there's an issue inside `kysely-planetscale` or `@planetscale/database` that's preventing the normal transaction semantics from working, or if there's a legitimate lock contention in PlanetScale / Vitess that is only now showing up for you because previously it would have silently failed.

**Comment by OultimoCoder** [comment]
Does the demo code not prove there is no legitimate lock contention? 

A failed transaction followed by a delete should always work, no?

I might have an investigate into this too then see if I can find out anymore.

**Comment by jacobwgillespie** [comment]
> A failed transaction followed by a delete should always work, no?

Possibly - how much traffic is this app getting? If you're getting this with a single request, then I'd expect it to work, but if there's concurrent requests / queries then it gets more complicated.

**Comment by OultimoCoder** [comment]
Tested locally. Repeatable and reproducible with a single request. 0 concurrency.

**Comment by OultimoCoder** [comment]
Ill write a reproducible script for this.

**Comment by OultimoCoder** [comment]
@jacobwgillespie hey!

I have a repo here which reproduces the issue pretty consistently. Very minimal code with a quick readme on how to run it (you just need a fresh planetscale db/branch).

https://github.com/OultimoCoder/kysely-planetscale-issue-20

**Comment by jacobwgillespie** [comment]
@OultimoCoder thank you for that reproduction, that's super helpful - I ran it with quite a bit of logging inside the `kysely-planetscale` code, and it appeared to be issuing the expected queries, it correctly created a separate connection for the transaction, etc.

I adapted it to use `@planetscale/database` directly, with raw SQL queries, and it's still timing out due to a lock. You can see the updated code here: https://github.com/jacobwgillespie/kysely-planetscale-issue-20.

So, I think this is either an issue inside `@planetscale/database`, or some misunderstanding of how locks are working with PlanetScale. I would assume the former, since it feels like a single transaction, with `ROLLBACK` shouldn't affect anything like you said.

**Comment by OultimoCoder** [comment]
Thanks so much for looking into this! Should I raise this issue on the `@planetscale/database` repo and point to both of our codebases?

**Comment by jacobwgillespie** [comment]
I think that makes sense, yeah, they're pretty responsive over there 👍

**Comment by jacobwgillespie** [comment]
This was fixed in `@planetscale/database` v1.10.0 🎉 

https://github.com/planetscale/database-js/releases/tag/v1.10.0
